Improve: add faint support and separate bold from colours

SlySven opened the pull request:
#### Brief overview of PR changes/additions
This disconnects the link that Mudlet had between ANSI colour numbers 8 to 15 (in the 16 and 256 colour ranges) from emboldening (increasing the "weight") of the Font used.

It also adds support for the ANSI Escape SGR sequence "2" option for "faint" (a.k.a. "dim") which provides two additional "weights" that can be combined with the bold effect.

So as to show the effects four `QLineEdit`s and `QLabel`s have been added beneath the font selection controls so that if the user types some text into the first (normal) one the text is reproduced in all of them but with the different combinations applied in each different box - there is also some placeholder text in each of them which will still partially show the effects even if nothing is entered into the first one.

#### Motivation for adding to Mudlet
I have often wondered about why we have been apply the bold font effect to characters that use the second eight of the ANSI 16 / 256 colour indexes - and, since it dates back to early Heiko era code 3e127f554bb23f8e86fbe64c560bccfc7bd096d0 it is hard to see as there is no documentation. My best *guess* is that it was intended to emphasis that the colour was from the second eight of the sixteen which for Muds that use that many colours are the brighter ones (with the RGB components at "full-strength") compared to the first eight which are less bright/intense.

#### Other info (issues closed, discussion etc)
This should help to close #3693.

I am considering providing (in a separate PR) a button on the display preferences for the colours of the main console that will "Reset to 8-color defaults" and to rename the existing one "Reset to 16/256 color defaults" which will swap (all but the "blacks") the first 8 with the second 8 so that MUDs that only use 8 colours will automagically show the "full-strength" colours for the eight that they do use.

SlySven committed:
Improve: add faint support and separate bold from colours

This disconnects the link that Mudlet had between ANSI colour numbers 8 to
15 (in the 16 and 256 colour ranges) from emboldening (increasing the
"weight") of the Font used.

It also adds support for the ANSI Escape SGR sequence "2" option for
"faint" (a.k.a. "dim") which provides two additional "weights" that can be
combined with the bold effect.

So as to show the effects four `QLineEdit`s and `QLabel`s have been added
beneath the font selection controls so that if the user types some text
into the first (normal) one the text is reproduced in all of them but with
the different combinations applied in each different box - there is also
some placeholder text in each of them which will still partially show the
effects even if nothing is entered into the first one.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

mpconley commented:
Would love to try this once MacOS is back.

SlySven committed:
Merge branch 'development' into Improve_addFaintSupportAndSeparatorBoldFromColors

mpconley commented:
<img width="996" alt="Screenshot 2023-11-13 at 6 48 21 PM" src="https://github.com/Mudlet/Mudlet/assets/3058178/e084d856-63a3-4f51-ade0-a10332efa279">

[crash.txt](https://github.com/Mudlet/Mudlet/files/13343028/crash.txt)

mpconley commented:
Before getting too discouraged, let's see if a PTB will work.  None has posted yet.

SlySven committed:
Merge branch 'development' into Improve_addFaintSupportAndSeparatorBoldFromColors

mpconley commented:
Yes, it does look like 'ansi' and 'xterm-256color' are returning bold now, but perhaps 'dim' or 'faint' is not faint yet.

SlySven committed:
Improve: add BoldIsBright setting

Restore some of the removed code that allowed the Bold `<SGR>1m` code to combine with
and select colors 8 to 15 when the basic `<SGR>30m` to `<SGR>37m` codes for colors 0
to 7 to specify a foreground color. This is a known "hack" that some terminals as
well as MUD servers use to get sixteen colors. As it interferes with other, more
advances ways of selecting not just 16 but also 256 or even 16M colors this behaviour
is now made optional with a setting I have called "Bold is bright" which defaults to
"ON" to reproduce, *mostly* Mudlet's previous behaviour for foreground (but not
background) colors. However unless the user is playing a MUD that does use only 16
colors and does it this way they probably will want to switch this option OFF in
practice.

A further commit to follow is intended to make it easier for the end-user to see
what types of color selection codes the Game Server is sending so as to guide them
in that choice.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Improve: add code to detect specific SGR codes in incoming data

Monitors the incoming Server data and records the use of some `SGR`
formatting codes so that their use can be displayed in a section of the
"Display colours" tab of the profile preferences. As well as the type of
codes used to select the foreground and background colours it also monitors
the use of the Bold (1), Faint (2), Blinking (5), Fast Blinking (6) and
Concealed (8) codes. The first two, along with the color ones will help
the user to deduce how many colours the MUD Game Server might be using and
thus whether to enable or disabled the, now optional since the previous
commit, use of the Bold attribute to select the second set of eight
foreground colours of sixteen which some 16 colours MUDs and some terminals
use.

So as to reduce screen clutter the status is displayed on a set of disabled
(as they are not user adjustable) checkboxes that are all hidden until the
end-user requests them to be shown by clicking on the groupbox that
contains them and the button to reset the detection should they have just
told the Game Server that they want a different colour setting. The status
displays are updated in real-time and will get a checkmark as soon as the
corresponding code is detected on the incoming data. So as to avoid
confusion with anything that `feedTrigger(...)` might insert into the data
such data is disregarded - which does mean that testing cannot be carried
out using that function.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Improve_addFaintSupportAndSeparatorBoldFromColors

mpconley commented:
On StickMUD, when I set "terminal ansi" and use the "colors" command:
* I am not seeing faint / dim resolve, and
* bold black (the line that almost can't be seen there) shows a color other than black for bold black

<img width="572" alt="Screenshot 2023-12-03 at 4 29 25 PM" src="https://github.com/Mudlet/Mudlet/assets/3058178/06dc81a6-b436-4999-b963-5e720ef0fe7c">

When I do "terminal xterm-256color" or "terminal truecolor" all is good still.

SlySven commented:
Have a look at which boxes are getting checked in the bottom of the revised "Display Colors" tab - by default (so as to reproduce prior behaviour) the "basic 8 color" ciodes (`<SGR>30m` to `<SGR>37m`) will be combined with Bold (`<SGR>1m`) to use the second eight colours - but if you uncheck the "BOLD is Bright" option then that black will stay black...

Here is a screen shot after I encountered a 16 Colour MUD that uses that option - which is a long-standing hack on some MUDs and terminals apparently:
![image](https://github.com/Mudlet/Mudlet/assets/6163092/5cf8e447-5f78-47a4-83db-77eeae949718)

mpconley commented:
I checked the box, and it did seem to bring me back to the default behavior. I restarted the client and the setting did not save. Also, other than the reset, are all of the checkboxes read-only? No dim/faint.  Thanks for working on this :)

SlySven commented:
> ...  I restarted the client and the setting did not save...

I'll have to check that, I did put in code to save it - but maybe I borked it...

> ... Also, other than the reset, are all of the checkboxes read-only? ....

Yeah, they are to indicate what has been seen in the incoming data - but there isn't a "read-only" option for `QCheckBox`es (unlike there is for `QLineEdit`s) so I had to disabled them to prevent any user interaction - apart from the tool-tips that some of them have. The non-colour related ones should help to debug when you might be getting codes that Mudlet does not completely handle **yet** :grinning: (blinking, concealed, alternate fonts) or which might only show up if there is enough weight variations available in the selected font family (faint).

SlySven committed:
Merge branch 'development' into Improve_addFaintSupportAndSeparatorBoldFromColors

SlySven commented:
> I'll have to check that, I did put in code to save it - but maybe I borked it...

I've just done so and have confirmed that it is being saved (and then loaded) correctly - if you want to check you can look in the Game save `.xml` file for the `BoldIsBright` attribute to the `Host` element:
![image](https://github.com/Mudlet/Mudlet/assets/6163092/c3be0cd2-b74c-4f6e-b691-b72a85052f68)

SlySven committed:
Merge branch 'development' into Improve_addFaintSupportAndSeparatorBoldFromColors

SlySven committed:
Fix: insert a line missed in a previous commit

It records that the SGR code for "faint" has been seen...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
> ... No dim/faint. ...

Yeah I just spotted that I had missed the line that records that that code had been received - now sorted.

vadi2 reviewed (commented):
Could you help me understand a bit more about how this option functions? I'm curious if the default of bold being bright remains unchanged, and whether using faint/bold fonts is an intentional choice users must make?

On another note, we're juggling quite a bit of content in the settings dialog lately! It's become a real puzzle fitting all these options, knobs, and dials in. 😅 I'm inclined to consider space efficiency here unless this feature garners significant interest. Lets remove it until there is interest?

![image](https://github.com/Mudlet/Mudlet/assets/110988/0b5ed091-c864-4d07-9cc3-43ccbcb8ddd8)

And could you clarify what the following option does?

![image](https://github.com/Mudlet/Mudlet/assets/110988/a2f75293-7d1f-4834-a38e-65fa69584709)

SlySven commented:
> Could you help me understand a bit more about how this option functions? I'm curious if the default of bold being bright remains unchanged, and whether using faint/bold fonts is an intentional choice users must make?

Up until this PR, the code put in by Heiko many years ago meant that the `<SGR>1m` code made the font **Bold** but it also caused the colour selected by the `<SGR>30m` to `<SGR>37m` codes to be the second 8 colours (which are brighter than the first 8 colours if they are set to their defaults). This PR removes this second action if the "Bold is bright" option is unchecked (but to reproduce the behaviour of prior Mudlet code it will be checked for new profiles). The reason for the four "sample" text display boxes under the font selection is because the Qt font selection code may not choose a font that has all four different font weights to produce a graduation in weight in the series "Faint" -> "Normal" -> "Faint + Bold" -> "Bold". The only way to be sure is to try a font and see if the four samples look different.

The "Bold is bright" option has no bearing on this - however it does mean that if enabled for a 16 colour MUD that does select the second 8 colours with this method then Bold text will (by default) be a ***different*** colour as well as the font being emboldened.

> And could you clarify what the following option does?

The GroupBox checkBox is - like the text says shows or hides the contents of the groupbox.

> ... I'm inclined to consider space efficiency here unless this feature garners significant interest. Lets remove it until there is interest? ...

And this is precisely to improve the space efficiency by hiding the details until the end-user wants it shown - the purpose of the checkBoxes is so that it is possible to determine which types of colour and other particular \<SGR>
codes the MUD Game Server is sending out so that both the end-user (and people trying to debug things) can better work out whether things are working and what options might be best to choose. Removing it would be counter productive as the End-user will not be able to tell what colour codes and other rarer codes (faint/blinking/fast blinking/concealed/alternate fonts) the Game Server is issuing...

@mpconley has already expressed an interest in this (for, I guess, "StickMUD"?)

mpconley commented:
There is a lot of interest in seeing BOLD + COLOR render as BOLD + COLOR on the screen. If I connect to my game with the "terminal" window on a Mac and put a character in ANSI mode on the game, BOLD + COLOR renders with ANSI mode. These are the defaults on a Mac terminal window that achieve that. Neither the user or the game admin does anything other than the user switching to ansi or 256 colors settings on the game. Hopeful that Mudlet would allow the same and perhaps these UI settings could inspire?

<img width="657" alt="Screenshot 2023-12-24 at 10 39 23 AM" src="https://github.com/Mudlet/Mudlet/assets/3058178/8e23f7d6-2452-4364-bc26-35af797b5f99">

vadi2 commented:
Really appreciate the work here! I'm in for unlocking the ability to use faint fonts, but it cannot be done at the expense of adding more visual complexity to the settings menu. 

After we solve that, we also need a clear strategy on how we introduce this option - that it, how can players or game admins make use of this.

vadi2 commented:
The mac terminal settings unfortunately don't help here - what does it mean not to use bold fonts or ANSI colors?

mpconley commented:
> The mac terminal settings unfortunately don't help here - what does it mean not to use bold fonts or ANSI colors?

With those minimal settings on the mac terminal (not a lot of visual complexity), users are able to tune their experience and get the desired effect of bold and colors. Will we need much more?

SlySven commented:
@mpconley When you say BOLD + COLOR - do you mean presenting the text with the **bold** effect AND with a *different, brighter colour*? A purpose of this PR was so that for MUDs that want **BOLD** text they can have that (or *faint* text) without any colour change (by unchecking the "BOLD is Bright" option). I hadn't, though, considered whether other users would want the \<SGR>1m (bold/bright) ANSI ESC code to ***only*** change the colour **without** modifying the text presentation. If that is also a consideration then a revision to this PR will be needed... :thinking:

vadi2 commented:
I think we should have a call about this, this would be too complex to navigate over text. I'm still not across how does https://github.com/Mudlet/Mudlet/pull/7010#issuecomment-1868545002 exactly work, and the options in the terminal menu aren't [that intuitive](https://github.com/Mudlet/Mudlet/pull/7010#issuecomment-1868546776).

mpconley commented:
Starting from square 1, gracefully presenting a welcome screen to a player that either has or does not have escape codes in it is important. Usually games will negotiate for a second or two to determine if they could add some color to the welcome page. With terminal_type, we'll likely get enough to determine if we could add "ansi". Being able to add the basic effects of bold underline strikethrough, etc. are what we need out of the box with the clients. That is what the most basic windows, mac, or linux telnet connections could do and I have displayed earlier in the PR comments. "use bold fonts", "display ansi colors".  That is the baseline and is much as games will start delivering to their connecting players.  **Bold + colors with basic "ansi" or "256colors" codes sent to Mudlet don't render as bold today.**  That's not just Tamarindo telling Mudlet this, it is a lot of game admins :) Recently we found out that Bold + colors with "truecolor" codes do work with Mudlet today and it generated a lot of excitement.

Mission #1 is to get bold and colors working with the "ansi" and "256colors" options without having to tune anything in the Mudlet UI.  Why?  If the most basic terminals can do it, Mudlet should be able to also.  When we solve that, then we could move on to enhancing the UI with "using bright colors for bold text" and everything else @SlySven  wants to do in this PR (which is a bit above my head).  Please ensure with no UI option changes we can have the basic bold as soon as games send "ansi" or "256colors" codes (again, that works with "truecolor" so we can do it, right?).

Keeping it simple for players and game admins is to negotiate to the optimal basic experience. Then the reason people pick Mudlet is the good things @SlySven is adding on top as icing on the cake. Let's just make sure we let them eat cake :)

vadi2 commented:
> Bold + colors with basic "ansi" or "256colors" codes sent to Mudlet don't render as bold today.

That's how every client works today, isn't it? I'm happy to consider a change to move things forward, but to establish a baseline here - to my knowledge Mudlet is following the trend.

SlySven commented:
To produce colour there are multiple schemes which I summarise thus:
1. Basic 8 (fore ground) colours (ANSI Colors 0 to 7) - this is done with the Set Graphics Rendition codes of `<SGR>30m` to `<SGR>37m`. This is the simplest, most basic system. Note however that these colours are defaulted in Mudlet to be "dark" in that the red, green and blue components are only at 128/255 of full intensity
2. Basic 16 (fore ground colours (ANSI Colors 0 to 15)  - this is done with the Set Graphics Rendition codes of `<SGR>30m` to `<SGR>37m` for the first 8 colours **and then combined with** the `<SGR>1m` (Bold/Bright) to "shift" them into the second 8 colours. This is the cause of the kludge that I am trying to resolve, supporting this mode is what is Mudlet has always done in the past but that `<SGR>1m` (Bold/Bright) code can/is also used to modify the appearance of the text.
3. Basic 16 (fore ground) colours (ANSI Colors 0 to 15)  - this is done with the Set Graphics Rendition codes of `<SGR>30m` to `<SGR>37m` for the first 8 colours and `<SGR>90m` to `<SGR>97m` for the second 8 colours  - TBH this is a better option than 2. because it does not interfere with the Bold code which can then be used solely to adjust the way the font is rendered.
4. 256 (fore ground) colours (ANSI Colour 0 to 255 ) - this is done with `<SGR>38;5;0m` to `<SGR>38;5;255m` (or  `<SGR>38:5:0m` to `<SGR>38:5:255m`)
5. 16M "Truecolor" (fore ground) direct colours - this is done with `<SGR>38;2;rrr;ggg;bbb...m` or `<SGR>38:2:rrr:ggg:bbb...m` or  `<SGR>38:2::rrr:ggg:bbb...m`

So, as a MUD Server, if you are aware that the Client can do 256 or TRUE colour you do not need to worry about the BOLD effect having any effect on the colours used for the text and that it will only affect the way the text is rendered. However if you only know it can do basic colour (ANSI colour I suppose) you will have to restrict yourself to 8 colours (and be aware that those will be dark colours) - or ask the user to try them in order of 3, 2 and 1 to see which ones work for them...

As an End-user, if the MUD you connect to uses any of the **first 3** you will need to try and find out more to get the best effects from the Server.

Note that I only talk about foreground colours because although 1 and 3 (like 4 and 5) have completely separate specifiers for the background colour, case 2. queers things because it is not possible to be sure what the background colour should be if the bold effect to shift to the "brighter" colours is to be applied to just the foreground colour or both of them.

mpconley commented:
We are aligned now on bold after re-running tests. Looking at the defaults in a side by side comparison of current Mudlet vs. this PR, bold works for ANSI and 256 colors, which satisfies what the community wants to see. Faint or dim does not seem to be changed to me.  Here is what we send for faint / dim --> `\x1B[2m`

* **ansi** - Using ANSI codes, current Mudlet (left) is not bold, this PR (right) is bold.
<img width="1217" alt="Screenshot 2023-12-26 at 1 32 41 PM" src="https://github.com/Mudlet/Mudlet/assets/3058178/e6e9c75b-61bd-4baa-aa6d-67c5eea3133d">

* **256color** - Using 256color codes, current Mudlet (left) is not bold, this PR (right) is bold.
<img width="1213" alt="Screenshot 2023-12-26 at 1 33 40 PM" src="https://github.com/Mudlet/Mudlet/assets/3058178/7b469016-0955-4934-8065-ccb2b7f19feb">

* **truecolor** - Using truecolor codes, current Mudlet (left) is bold (as expected), this PR (right) is bold (as expected).
<img width="1201" alt="Screenshot 2023-12-26 at 1 34 21 PM" src="https://github.com/Mudlet/Mudlet/assets/3058178/5cadd35e-af47-4c9a-a9f2-4095b931f72f">

SlySven commented:
For the first of those, can you show me the "Colors" tab on the preferences with the (new) "SGR code detection" box checked (so that it is visible) - ideally if you can click the "reset" button at the bottom (to clear all those read-only checkboxes) after you enter `terminal ansi` and just before you enter `colors` - and if you can do that for the new "Bold is Bright" checkbox being unchecked and checked.

I would expect that the screen shot you are showing for the "ANSI" color set is with the "Bold is Bright" checkbox checked so that the bold code is switching to the "dark-grey" (ANSI Colour 8) instead of being "black" (ANSI Colour 0) - this is what will happen in the current "Release" version. If you want the black text to be left at black you will need to uncheck that option - but whether you then get 16 different colours remains to be seen and will depend on whether the StickMUD server does 16 colours using method 2 or 3 in the above list...

mpconley commented:
Starts out like this:

<img width="1518" alt="Screenshot 2023-12-26 at 2 29 22 PM" src="https://github.com/Mudlet/Mudlet/assets/3058178/18cbe64c-a07f-40f1-8c21-9aa3a6613d36">

And here is how it looks with the bold is bright unchecked:
<img width="1505" alt="Screenshot 2023-12-26 at 2 22 28 PM" src="https://github.com/Mudlet/Mudlet/assets/3058178/c953e0e3-1619-43e7-bf19-7523308eb02f">

The reset button did not do anything for me either time, but I also did not change anything other than bold is bright.

SlySven commented:
As for the Dim/Faint - what is the Font being used - and does your system have more weights other than "Font Name" (or "Font Name Regular" or "Font Name Book") and "Font Name Bold"? If it only has those then, yeah, you probably won't get faint working, for instance compare the Bitstream Vera Sans Mono (actually the Mudlet default) - which only has 2 weights (Regular and Bold):

![image](https://github.com/Mudlet/Mudlet/assets/6163092/4dbce62c-f414-4ded-b9ef-4cc933e3a501)

with Fira Code which has 5 weights (looks like I have a duplicate) - and which thus ***can*** provide the dim/faint (and the demi-bold) font needed:

![image](https://github.com/Mudlet/Mudlet/assets/6163092/67523b81-2431-4f3b-8850-ec2a485a654e)

SlySven commented:
> The reset button did not do anything for me either time, but I also did not change anything other than bold is bright.

It does, but you will need to check that bottom group box to see anything - the checkboxes in there get checked (in real-time BTW) when the corresponding SGR code is seen - that reset button clears them all...

Edit: perhaps it would be more sensible to hide it as well when the group box is unchecked (disabled). :thinking:

mpconley commented on src/XMLimport.cpp:
```diff
@@ -957,6 +957,13 @@ void XMLimport::readHost(Host* pHost)
         pHost->setLargeAreaExitArrows(false);
     }
 
+    if (attributes().hasAttribute(qsl("BoldIsBright"))) {
```
When this merges, there was some work done in PR #7036 to use code like this you may want to consider: `setBoolAttributeWithDefault(qsl("BoldIsBright"), pHost->mBoldIsBright, true);`.  It won't show as a conflict, I think.

SlySven replied:
As that has been merged I may get something to do when I try to click on "Update branch". :grinning:

SlySven commented:
With regard to fonts that are Monospace and that have multiple weights (enough to show a faint version) I made an error before in quoting "Fira Code" as a suitable font - in that although it "looks" to be monospaced it does not provoke a `true` result from `QFontInfo::fixedPitch(QFont("Fira Code"))` - in fact I should have said "Fira Mono" but that only seems to have three weights available.

OTOH some hunting has found the following that do seem to have a number of weights and which are mono-spaced and do seem to show the effect:
* Cascadia Code
* Cascadia Code PL
* Cascadia Mono
* Cascadia Mono PL
* IBM Plex Mono (has more than 4 weights all with separate italics as well)
* JetBrains Mono (has more than 4 weights all with separate italics as well)
* League Mono (has more than 5 weights all with 5 different letter spacings - though not italics)

SlySven committed:
Merge branch 'development' into Improve_addFaintSupportAndSeparatorBoldFromColors

SlySven committed:
Merge branch 'development' into Improve_addFaintSupportAndSeparatorBoldFromColors

Conflicts resolved in:
* src/Host.h
* src/TLuaInterpreter.cpp (some new parts now in src/TLuaInterpreterUI.cpp)

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: hide by default extra font samples & SGR code detector reset button

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Improve_addFaintSupportAndSeparatorBoldFromColors

SlySven committed:
Revise: add BoldIsBright option to get/setConfig(...) Lua API functionality

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
Anyone know the answer to https://github.com/Mudlet/Mudlet/pull/7010#issuecomment-1869067967 ?

atari2600tim commented:
The screenshot in [#3693 ](https://github.com/Mudlet/Mudlet/issues/3693#issuecomment-923402890) uses game code of
```
  %^RED%^ABC%^BOLD%^DEF%^RESET%^GHI
         ABC%^BOLD%^DEF%^RESET%^GHI
%^WHITE%^ABC%^BOLD%^DEF%^RESET%^GHI
```
to display three lines starting out differently with red/uncolored/white ABC then add bold and DEF, then a reset code (style and color), then GHI.  It uses a table similar to [this one](https://github.com/fluffos/lima/blob/dbcef2a9b2919474145e6a533fc434d67f6c7f5a/lib/data/daemons/ansi.o#L2) so RED is `[31m`, BOLD is `[1m`, RESET is `[0m`, WHITE is `[37m`.

If you look at the second line second set of characters, DEF, with default coloring and only bold added, it looks like Mudlet is the only one that makes the characters thicker.  With red and white, Mudlet with color+bold changes the color just like all the others.  Uncolored default with bolded added is what shows as bold.  It isn't super obvious, but look specifically at the space inside of the `D` since both sides are pushing in on that space. 
![image](https://github.com/Mudlet/Mudlet/assets/29287358/b19b4d2d-f308-4652-acd6-83c5d6e1c84a)

I'd have to read through everything to comment on anything else, only remembered taking this screenshot.  But yes, color+bold is already consistent with others, it's default+bold where it differs currently.

SlySven commented:
> Anyone know the answer to [#7010 (comment)](https://github.com/Mudlet/Mudlet/pull/7010#issuecomment-1869067967) ?

When one says Bold with ANSI what is one saying though? Is that just 8 ANSI colours or 16? And is that Bold as in **BOLD** or Bold as in Brighter colours? If one wants the brighter 8 colours from the 16 you either have to send the OSC commands to swap the first set of 8 colours with the second 8 - or, perhaps simpler, just use the 256 Colour codes for clients that have been identified as Mudlet...

vadi2 commented on src/Host.h:
```diff
@@ -706,6 +707,17 @@ class Host : public QObject
     // shortcut to switch between the input line and the main window
     CaretShortcut mCaretShortcut = CaretShortcut::None;
 
+    // Support a long-standing hack for using the <SGR>1m (Bold) code to select
```
```suggestion
    // Support a long-standing hack among all clients for using the <SGR>1m (Bold) code to select
```

SlySven replied:
Do we know that, or just suspect it?

vadi2 replied:
We know it per https://github.com/Mudlet/Mudlet/pull/7010#issuecomment-2094869387 which answered https://github.com/Mudlet/Mudlet/pull/7010#issuecomment-1869067967

SlySven commented:
@atari2600tim if you tried that with current development branch code and this one (with the "Bold is bright" option enabled as it will be by default) then I'd expect BOTH of them to have a bold **DEF** on all three lines. Unchecking that option in this PR's one will just cause the colour of the DEF to be the same as the ABC on all three lines (and the second and the third lines to be the same if the colours are those that the "Reset" button on the "Colour view" tab {if I've remembered the name correctly} will set them to). I think that will be different to the example you showed from that earlier #3693 ...

vadi2 commented on src/dlgProfilePreferences.cpp:
```diff
@@ -4517,3 +4548,106 @@ void dlgProfilePreferences::slot_changeLargeAreaExitArrows(const bool state)
+}
+
+// The weights used need to match those in TTextEdit::drawGraphemeForeground(...)
+void dlgProfilePreferences::updateFontSampleDisplays(const QFont& newFont)
```
The font preview might be handy for development, but I don't foresee the majority of players needing this option, and as the settings menu is already super cramped and way too busy, we shouldn't add this feature. There are dedicated font preview tools that do a better job already 🙂

SlySven replied:
The end-user is likely going to need it to check whether the font they have chosen has the additional weights to provide distinct faint and demi-bold rendering, and that they have selected the correct (the "regular" or normal weight on systems that list all the weight variations as separate fonts). Sadly, most *don't* provide the former and in the latter case it is easy to, say, click "DejaVu Sans Mono Bold" when one really ought to choose "DejaVu Sans Mono". As a concession to reduce the space needed I have now hidden the other weight displays unless requested but they do really help to verify that the extra weights do show up differently. A separate font preview tool is not going to necessarily show the particular font weights that have been hard coded into Mudlet (see `./src/TBuffer.h` from line 266).

vadi2 replied:
Games also offer text previews. Should there be a game that implements faint text (do we actually know of any?) they surely will have a way for the player to know if it's working for them. It is important to keep complexity of Mudlet low; more knobs != better client.

SlySven replied:
Okay - under protest - I've whipped out all the extra bits try 0611553b0928b8f38ddbc49dd9c472b4f8bb7d95 now... 😢

vadi2 commented on src/ui/profile_preferences.ui:
```diff
@@ -1916,6 +2059,192 @@ you can use it but there could be issues with aligning columns of text</string>
+       <item>
+        <widget class="QGroupBox" name="groupBox_sgrCodesSeen">
+         <property name="title">
+          <string>&quot;Set Graphics Rendition&quot; codes detected (check this to show details)</string>
```
The settings' menu purpose for working with settings and I'm not certain diluting it to do debugging as well is a good idea.

SlySven replied:
It is the most appropriate place that I can think of to implement it - if you can see from that that you are only getting 8 ANSI colours - like you do from, say, WotMUD, then you can work out that you probably want (like I do) to swap the first set of 8 colours with the second 8. Also, if you can see that your MUD is sending you AXTerm colour codes to do the second half of the 16 Colour set then you can reasonable uncheck the "BOLD-is-Bright" option) - as you can if it is sending 256 or 16M colour codes. Anyhow I think it will help us and others debug problems when end-users complain about their MUD not showing the colours correctly.

Knowing that you are seeing Alternate font/blinking/concealed codes that Mudlet does not currently render will help us and MUD Server operators develop these features - should we/they wish to in the future.

vadi2 replied:
I think this is addressing a problem that we think exits in theory but doesn't actually happen in practice. I haven't seen players struggle with this, but maybe I missed something? 

This makes the case for diluting what settings is about not very strong 😕

At most, this can perhaps be worked into the text analyser some way - even though the text analyser looks at one character at a time and this tool looks at all the text that as been received from the last reset.

SlySven replied:
Okay - under protest - I've whipped out all the extra bits try 0611553b0928b8f38ddbc49dd9c472b4f8bb7d95 now... :cry:

atari2600tim commented:
I'm eyeballing screenshots of the default settings, so if there is a font that looks same bolded or not, then might want to experiment.  I vaguely recall Fixedsys might be that way.

SlySven committed:
Revise: excise all the shiny extra UI goodness from the PR as requested

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Since there is no longer a preview one will have to use something like:
```lua
feedTriggers("\027[2mFaint \027[0mNormal \027[1;2mDemi-bold \027[22;1mBold\027[0m\n")
```
to compare the various weights to see if the font you choose works with faint and bold.
![Screenshot_20240507_212626](https://github.com/Mudlet/Mudlet/assets/6163092/993610db-86b4-4038-8b1f-b2f4bffb7f04)
![Screenshot_20240507_212711](https://github.com/Mudlet/Mudlet/assets/6163092/7979b3a8-7f64-467f-bca0-4e2b0edc7065)
![Screenshot_20240507_212742](https://github.com/Mudlet/Mudlet/assets/6163092/795a9b47-e66f-4705-8103-73b565029042)

vadi2 commented:
I think that works, one could put a Mudlet package together that shows you such previews :)

Players who would be interested could install it off the forums or wiki and this opt into this functionality.

vadi2 reviewed (approved):
Thanks for adding this feature in, especially in a way that doesn't complicate Mudlet for players 👍
